src/toxcore/dht_new/codec.rs

```diff
@@ -0,0 +1,343 @@
+/*
+    Copyright (C) 2013 Tox project All Rights Reserved.
+    Copyright © 2016-2017 Zetok Zalbavar <zexavexxe@gmail.com>
+    Copyright © 2018 Namsoo CHO <nscho66@gmail.com>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*! Codec for encoding/decoding DHT Packets & DHT Request packets using tokio-io
+*/
+
+use toxcore::dht_new::packet::*;
+use toxcore::tcp::secure::*;
+use toxcore::dht_new::binary_io::*;
+
+use nom::Offset;
+use std::io::{Error, ErrorKind};
+use bytes::BytesMut;
+use tokio_io::codec::{Decoder, Encoder};
+
+/// implements tokio-io's Decoder and Encoder to deal with DHT Packet
+pub struct DhtPacketCodec {
+    channel: Channel
+}
+
+impl DhtPacketCodec {
+    /// create a new DhtPacketCodec with the given Channel
+    pub fn new(channel: Channel) -> DhtPacketCodec {
+        DhtPacketCodec { channel: channel }
+    }
+}
+
+const DHT_PACKET_HEADER_SIZE: usize = 57;
+const MAX_DHT_PACKET_SIZE: usize = 512;
+
+impl Decoder for DhtPacketCodec {
```

Comment:
You implemented Decoder for TCP. There is another UDPCodec which must be implemented since DHT packets are udp only